sbt-metals/src/main/scala/scala/meta/metals/MetalsPlugin.scala

```diff
@@ -44,4 +71,89 @@ object MetalsPlugin extends AutoPlugin {
   def isScala3: Def.Initialize[Boolean] = Def.setting {
     ScalaInstance.isDotty(scalaVersion.value)
   }
+
+  def javaSemanticdbOptions: Def.Initialize[Seq[String]] = {
+    def exportsFlags(version: JdkVersion): List[String] = {
+      if (version.major >= 17) {
+        val compilerPackages = List(
+          "com.sun.tools.javac.api", "com.sun.tools.javac.code",
+          "com.sun.tools.javac.model", "com.sun.tools.javac.tree",
+          "com.sun.tools.javac.util"
+        )
+        compilerPackages.flatMap(pkg =>
+          List(s"-J--add-exports", s"-Jjdk.compiler/$pkg=ALL-UNNAMED")
+        )
+      } else Nil
+    }
+    Def.setting {
+      val defined = javaHome.value
+      val value = defined.getOrElse(file(System.getProperty("java.home")))
+      JdkVersion.getJavaVersionFromJavaHome(value) match {
+        case None => Seq.empty
+        case Some(version) =>
+          val sourceRoot = (ThisBuild / baseDirectory).value
+          val targetRoot = (Compile / semanticdbTargetRoot).value
+          val pluginOption =
+            s"-Xplugin:semanticdb -sourceroot:${sourceRoot} -targetroot:${targetRoot} -build-tool:sbt"
+          pluginOption :: exportsFlags(version)
+      }
+    }
+  }
+
+  case class JdkVersion(major: Int) {
+
+    def hasJigsaw: Boolean = major >= 9
+  }
```

Comment:
copypasted it from `JavaIteractiveSemanticdb`.
Technically it's possible to extract into a module and share it for `metals` and `sbt-metals` but I'm not sure if it's worth it.